### DISCUSSION TITLE
understanding blackhole project

### DISCUSSION THREAD

**Discussion by ghost** [ExistentialAudio/BlackHole#310] [Q&A] [answered]
Ok, I want to ask few things about the projects. It's a very interesting and clean project and I want to contribute to it

1) Where's the main function in .c file? 
2) Is this a kind of Audio Driver? If yes can you suggest me where to write a sample driver for macOS? To start something from scratch and understand it clearly 
3) Where do I learn core-audio and where do I being with? Want to create something very basic 

Thanks. I love this project and want to maintain it for users/community for free. Earlier I was using soundflower :)

**Comment by devinroth** [comment, marked as answer]
There is no main.c.  It is an audio driver. Specifically a HAL plugin. Apple is notorious for lacking documentation on this stuff. Reading the headers is the best documentation for this stuff.

This is what's on the Apple Docs might be a good place to start. 

https://developer.apple.com/library/archive/documentation/MusicAudio/Conceptual/CoreAudioOverview/WhatisCoreAudio/WhatisCoreAudio.html

There's also a book called understanding CoreAudio.

This stuff is hard to understand. Be patient and keep at it.
